graphql/client.go

```diff
@@ -239,3 +258,161 @@ func (c *client) createGetRequest(req *Request) (*http.Request, error) {
+
+// recursively find all the fields that are type Upload.
+func findFiles(parentKey string, v reflect.Value, depth int) ([]*fileVariable, error) {
+	if depth > 10000 {
+		return nil, errors.New("possible stack overflow error")
+	}
+	fileVariables := []*fileVariable{}
+
+	if v.Kind() == reflect.Ptr {
+		if v.IsNil() {
+			return nil, nil
+		}
+		v = v.Elem()
+	}
+
+	if v.Type() == reflect.TypeOf(Upload{}) {
+		file := v.Interface().(Upload)
+		if file.Body == nil {
+			return nil, errors.New("Upload file body cannot be nil")
+		}
+		fileVariables = append(fileVariables, &fileVariable{
+			mapKey: parentKey,
+			file:   file,
+		})
+		return fileVariables, nil
+	}
+
+	switch v.Kind() {
+	case reflect.Struct:
+		for i := 0; i < v.NumField(); i++ {
+			field := v.Field(i)
+			fieldName := v.Type().Field(i).Name
+			jsonTag := v.Type().Field(i).Tag.Get("json")
+			if jsonTag != "" && jsonTag != "-" {
+				fieldName = jsonTag
+			}
+			files, err := findFiles(parentKey+"."+fieldName, field, depth+1)
+			if err != nil {
+				return nil, err
+			}
+			fileVariables = append(fileVariables, files...)
+		}
+	case reflect.Slice, reflect.Array:
+		for i := 0; i < v.Len(); i++ {
+			elem := v.Index(i)
+			files, err := findFiles(parentKey+"."+strconv.Itoa(i), elem, depth+1)
+			if err != nil {
+				return nil, err
+			}
+			fileVariables = append(fileVariables, files...)
+		}
+	}
+
+	return fileVariables, nil
+}
+
+func createUploadFileRequest(req *Request, endpoint string, fileVariables []*fileVariable) (*http.Request, error) {
+	httpRequest, err := http.NewRequest(http.MethodPost, endpoint, http.NoBody)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+	bodyBuf := &bytes.Buffer{}
+	bodyWriter := multipart.NewWriter(bodyBuf)
+	defer bodyWriter.Close()
+
+	// operations
+	requestBody, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling request: %w", err)
+	}
+	err = bodyWriter.WriteField("operations", string(requestBody))
+	if err != nil {
+		return nil, fmt.Errorf("error writing operations to body: %w", err)
+	}
+
+	// map
+	mapData := ""
+	// group files to avoid uploading duplicated files
+	filesGroup := [][]*fileVariable{}
+	for _, file := range fileVariables {
+		foundDuplicated := false
+		for group, fileGroup := range filesGroup {
+			file2 := fileGroup[0]
+			if file.file.FileName == file2.file.FileName {
+				f1, err := io.ReadAll(file.file.Body)
+				if err != nil {
+					return nil, fmt.Errorf("error reading file: %w", err)
+				}
+				f2, err := io.ReadAll(file2.file.Body)
+				if err != nil {
+					return nil, fmt.Errorf("error reading file: %w", err)
+				}
+				file.file.Body = bytes.NewReader(f1)
+				file2.file.Body = bytes.NewReader(f2)
+				if bytes.Equal(f1, f2) {
+					foundDuplicated = true
+					filesGroup[group] = append(filesGroup[group], file)
+					break
+				}
+			}
+		}
+		if !foundDuplicated {
+			filesGroup = append(filesGroup, []*fileVariable{file})
+		}
+	}
```

Comment:
No need to optimize duplicate files. Just join them into variables string.

Reply:
fix